go/test/endtoend/vtgate/gen4/gen4_test.go

```diff
@@ -276,3 +277,55 @@ func TestMultiColumnVindex(t *testing.T) {
 		})
 	}
 }
+
+func TestFanoutVindex(t *testing.T) {
+	ctx := context.Background()
+	conn, err := mysql.Connect(ctx, &vtParams)
+	require.NoError(t, err)
+	defer conn.Close()
+
+	tcases := []struct {
+		regionID int
+		exp      string
+	}{{
+		regionID: 24,
+		exp:      `[[INT64(24) INT64(1) VARCHAR("shard--19a0")]]`,
+	}, {
+		regionID: 25,
+		exp:      `[[INT64(25) INT64(2) VARCHAR("shard--19a0")] [INT64(25) INT64(7) VARCHAR("shard-19a0-20")]]`,
+	}, {
+		regionID: 31,
+		exp:      `[[INT64(31) INT64(8) VARCHAR("shard-19a0-20")]]`,
+	}, {
+		regionID: 32,
+		exp:      `[[INT64(32) INT64(14) VARCHAR("shard-20-20c0")] [INT64(32) INT64(19) VARCHAR("shard-20c0-")]]`,
+	}, {
+		regionID: 33,
+		exp:      `[[INT64(33) INT64(20) VARCHAR("shard-20c0-")]]`,
+	}}
+
+	defer utils.ExecAllowError(t, conn, `delete from region_tbl`)
+	uid := 1
+	// insert data in all shards to know where the query fan-out
+	for _, s := range shardedKsShards {
+		utils.Exec(t, conn, fmt.Sprintf("use `%s:%s`", shardedKs, s))
+		for _, tcase := range tcases {
+			utils.Exec(t, conn, fmt.Sprintf("insert into region_tbl(rg,uid,msg) values(%d,%d,'shard-%s')", tcase.regionID, uid, s))
+			uid++
+		}
+	}
+
+	newConn, err := mysql.Connect(ctx, &vtParams)
+	require.NoError(t, err)
+	defer newConn.Close()
+
+	for _, workload := range []string{"olap", "oltp"} {
+		utils.Exec(t, newConn, fmt.Sprintf(`set workload = %s`, workload))
+		for _, tcase := range tcases {
+			t.Run(workload+strconv.Itoa(tcase.regionID), func(t *testing.T) {
+				sql := fmt.Sprintf("select rg, uid, msg from region_tbl where rg = %d order by uid", tcase.regionID)
+				assert.Equal(t, tcase.exp, fmt.Sprintf("%v", utils.Exec(t, newConn, sql).Rows))
+			})
+		}
+	}
```

Comment:
If I understand correctly, we are inserting rows in all the shards for a given regionID, and we are checking which all shard will be reached by a `select *` query. Love this test 😮

Reply:
that is correct. :)